firebase-firestore/src/commonMain/kotlin/dev/gitlive/firebase/firestore/TimestampSerializer.kt

```diff
@@ -1,83 +1,66 @@
 package dev.gitlive.firebase.firestore
 
-import dev.gitlive.firebase.FirebaseCompositeDecoder
-import dev.gitlive.firebase.FirebaseCompositeEncoder
+import dev.gitlive.firebase.FirebaseDecoder
+import dev.gitlive.firebase.FirebaseEncoder
 import kotlinx.serialization.KSerializer
 import kotlinx.serialization.SerializationException
-import kotlinx.serialization.descriptors.SerialDescriptor
-import kotlinx.serialization.descriptors.StructureKind
+import kotlinx.serialization.descriptors.element
 import kotlinx.serialization.encoding.Decoder
 import kotlinx.serialization.encoding.Encoder
 
-abstract class AbstractTimestampSerializer<T> : KSerializer<T> {
+sealed class AbstractTimestampSerializer<T>(
+    private val isNullable: Boolean
+) : KSerializer<T> {
 
-    override val descriptor = object : SerialDescriptor {
-        val keys = listOf("seconds", "nanoseconds")
-        override val kind = StructureKind.OBJECT
-        override val serialName = "Timestamp"
-        override val elementsCount get() = 2
-        override fun getElementIndex(name: String) = keys.indexOf(name)
-        override fun getElementName(index: Int) = keys[index]
-        override fun getElementAnnotations(index: Int) = emptyList<Annotation>()
-        override fun getElementDescriptor(index: Int) = throw NotImplementedError()
-        override fun isElementOptional(index: Int) = false
+    override val descriptor = buildClassSerialDescriptor("Timestamp", isNullable = isNullable) {
+        isNullable = this@AbstractTimestampSerializer.isNullable
+        element<Long>("seconds")
+        element<Int>("nanoseconds")
     }
 
-    abstract override fun serialize(encoder: Encoder, value: T)
-    abstract override fun deserialize(decoder: Decoder): T
-
-    fun encode(encoder: Encoder, value: FirebaseTimestamp?) {
-        if (value == null) {
-            encoder.encodeNull()
-            return
-        }
-        val objectEncoder = encoder.beginStructure(descriptor) as FirebaseCompositeEncoder
-        when (value) {
-            is FirebaseTimestamp.Value -> {
-                objectEncoder.encodeObject(descriptor, 0, value.value)
-            }
-            is FirebaseTimestamp.ServerValue -> {
-                objectEncoder.encodeObject(descriptor, 0, FieldValue.serverTimestamp())
-            }
-            is FirebaseTimestamp.ServerDelete -> {
-                objectEncoder.encodeObject(descriptor, 0, FieldValue.delete)
+    protected fun encode(encoder: Encoder, value: FirebaseTimestamp?) {
+        if (encoder is FirebaseEncoder) {
+            encoder.value = value?.let {
+                when(value) {
+                    is FirebaseTimestamp.Value -> value.value
+                    is FirebaseTimestamp.ServerValue -> FieldValue.serverTimestamp()
+                    is FirebaseTimestamp.ServerDelete -> FieldValue.delete
```

Comment:
`ServerValue` is a valid case, but `ServerDelete` shall be removed and done in a generic way

Reply:
A more generic way you mean for all "special" classes?

Reply:
@corrado4eyes I meant calling `DocumentReference.update(FieldPath("field-containing-a-reference") to FieldValue.delete)` instead

Reply:
Yeah it's weird timestamps have a special delete value when any field can have a delete value.